tests/Fix64Tests.cs

```diff
@@ -361,6 +361,78 @@ public void Sqrt()
             }
         }
 
+        [Fact]
+        public void Ln()
+        {
+            for (int j = 0; j < m_testCases.Length; ++j)
+            {
+                var b = Fix64.FromRaw(m_testCases[j]);
+
+                if (b <= Fix64.Zero)
+                    continue;
+
+                if (b > Fix64.MaxValue / (Fix64)100)
+                    continue;
+
+                // Reduced precision requirements for very small values
+                double maxDelta = b < (Fix64)0.000001m ? 0.1 : (double)(Fix64.Precision * 10);
+
+                var expected = Math.Log((double)b);
+                var actual = (double)Fix64.Ln(b);
+                var delta = Math.Abs(expected - actual);
+
+                Assert.True(delta <= maxDelta, string.Format("Ln({0}) = expected {1} but got {2}", b, expected, actual));
+            }
+        }
+
+        [Fact]
+        public void Exp()
+        {
+            double maxDelta = (double)(Fix64.Precision * 10);
+            for (int i = 0; i < m_testCases.Length; ++i)
+            {
+                var e = Fix64.FromRaw(m_testCases[i]);
+
+                var expected = Math.Min(Math.Exp((double)e), (double)Fix64.MaxValue);
+                var actual = (double)Fix64.Exp(e);
+                var delta = Math.Abs(expected - actual);
+
+                Assert.True(delta <= maxDelta, string.Format("Exp({0}) = expected {1} but got {2}", e, expected, actual));
+            }
+        }
+
+        [Fact]
+        public void Pow()
+        {
+            for (int i = 0; i < m_testCases.Length; ++i)
+            {
+                var b = Fix64.FromRaw(m_testCases[i]);
+
+                if (b < Fix64.Zero)
+                    continue;
+                if (b > Fix64.MaxValue / (Fix64)100)
```

Comment:
For very large and very small bases the precision becomes bad or terrible, mostly because Ln isn't very precise anymore.
Example:
`Pow(0.0000000002328306436538696289, -0.5) = expected 65536 but got 65901.3926025061`
Or somewhat pathological cases like this where it really falls apart:
`Pow(2147483647.9999999997671693563, 0.9999999997671693563461303711) = expected 2147483637.25622 but got 1426713511.50211`

Finally, maybe the error should be measured relatively, instead of absolutely? 
`Pow(2147483647.9999999997671693563, 0.7911919844336807727813720703) = expected 24173977.2437025 but got 24173942.9738885`

I don't see a way to fully fix these cases, but also describing the exact limitations is somewhat difficult. I have refactored the test somewhat and am looking forward to your comments.

Reply:
So, the test skips half the range [0..MaxValue], and in the range ]1.. MaxValue/2] the delta is artificially reduced to a fraction of itself. I'm not sure I understand the rationale behind this? In any case I'm not comfortable with an operation that's effectively untested over a large range of inputs and boundary conditions (even though I'm guilty of that myself e.g. Tan - I do document that it's not tested well though).

If Ln is the culprit then let's find a more accurate Ln. You got me reading on the topic :) I think an accurate base-2 logarithm for a 64-bit integer would be a good start. https://graphics.stanford.edu/~seander/bithacks.html#IntegerLog has ideas, https://stackoverflow.com/questions/11376288/fast-computing-of-log2-for-64-bit-integers as well. 

Once we have that, Fix64 is just a long divided by 2^32, and log(x/y) is log(x) - log(y), so log2 of a Fix64 is log2(RawValue) - 32 (right?)

Once we have an accurate log2, ln x is just log2 x/log2 e.